src/main/java/org/openhab/tools/analysis/checkstyle/readme/MarkDownVisitor.java

```diff
@@ -0,0 +1,336 @@
+     */
+    private int currentLinePointer = 0;
+
+    /**
+     * A callback is used in order to use the protected methods of {@link AbstractStaticCheck}
+     */
+    private MarkdownVisitorCallback callBack;
+
+    private File file;
+    private List<String> lines;
+
+    public MarkDownVisitor(MarkdownVisitorCallback callBack, File file, List<String> lines) {
+        this.callBack = callBack;
+        this.file = file;
+        this.lines = lines;
+    }
+
+    /**
+     * Example of heading: #HomeMatic Binding
+     */
+    @Override
+    public void visit(Heading heading) {
+        String headerContent = getLiteralOfElement(heading);
+        validateHeader(headerContent);
+    }
+
+    /**
+     * This method processes the Headings in the README which the MarkDown
+     * parser returns.
+     * 
+     * @param headerValue - a String containing the literal of found Heading node
+     */
+    private void validateHeader(String headerValue) {
+        if (headerValue != null) {
+            int headerLineNumber = 0;
+            headerLineNumber = callBack.findLineNumber(lines, headerValue, currentLinePointer);
+            boolean isInvalidHeaderValue = headerLineNumber == -1;
+            if (isInvalidHeaderValue) {
+                logger.error("A header cannot be processed properly: {}", headerValue);
+            } else {
+                currentLinePointer = headerLineNumber;
+                boolean isHeaderAtEndOfFile = headerLineNumber == lines.size();
+                if (isHeaderAtEndOfFile) {
+                    callBack.log(headerLineNumber, EMPTY_LINE_AFTER_HEADER_MSG);
+                } else {
+                    boolean isNextLineEmpty = lines.get(headerLineNumber).trim().isEmpty();
+                    if (!isNextLineEmpty) {
+                        callBack.log(headerLineNumber, EMPTY_LINE_AFTER_HEADER_MSG);
+                    }
+                }
+            }
+        } else {
+            logger.warn("Occured an error while processing the md file {}. The header value is null.",
+                    file.getAbsolutePath());
+        }
+    }
+
+    /**
+     * A fencedCodeBlock is a text-formatted like a programming code, example:
+     * <code>private String name=null;</code>
+     */
+    @Override
+    public void visit(FencedCodeBlock fencedCodeBlock) {
+        String codeLiteral = fencedCodeBlock.getLiteral();
+        validateCodeSection(codeLiteral);
+    }
+
+    /**
+     * Processes FencedCodeBlock nodes the parser returns.
+     * 
+     * @param codeBlock - multiline String containing the literal of the code block
+     */
+    private void validateCodeSection(String codeBlock) {
+        int codeStartingLineNumber = 0;
+        int codeEndingLineNumber = 0;
+
+        // Splitting the String in case it is multiline, else the array would
+        // have only one element
+        String codeSectionLines[] = codeBlock.split("\\r?\\n");
```

Comment:
What warning do you get? I think the solution `\\r?\\n?` could given an error because it also matches nothing. So it would be come something else like `(\\r?\\n)|(\\r)` not as readable...